acurl/src/response.pyx

```diff
@@ -0,0 +1,224 @@
+#cython: language_level=3
+
+from libc.stdlib cimport free
+from cpython cimport array
+from json import loads
+from cpython.list cimport PyList_New
+from libc.stdio cimport printf
+
+cdef struct BufferNode:
+    size_t len
+    char *buffer
+    BufferNode *next
+
+cdef list acurl_extract_cookielist(CURL* curl):
+    cdef void* start_raw
+    cdef curl_slist *start
+    cdef curl_slist *node
+    cdef int len = 0
+    cdef int i = 0
+    acurl_easy_getinfo_voidptr(curl, CURLINFO_COOKIELIST, &start_raw)
+    start = <curl_slist*>start_raw
+    node = start
+    cdef lst = []
+    while node != NULL:
+        lst.append(node.data.decode("UTF-8"))
+        node = node.next
+    curl_slist_free_all(start)
+    return lst
+
+cdef class _Response:
+    cdef BufferNode* header_buffer
+    cdef BufferNode* header_buffer_tail
+    cdef BufferNode* body_buffer
+    cdef BufferNode* body_buffer_tail
+    cdef CURL* curl
+    cdef Session session
+    cdef object future
+    cdef readonly unsigned long start_time
+    cdef readonly Request request
+    cdef _Response _prev
+
+    def __cinit__(self):
+        # Technically it's dangerous to leave curl and session members
+        # uninitialized, but we hope no one calls our init method directly...
+        self.header_buffer = NULL
+        self.header_buffer_tail = NULL
+        self.body_buffer = NULL
+        self.body_buffer_tail = NULL
+        self._prev = None
+
+    def __dealloc__(self):
+        cdef BufferNode* ptr
+        cdef BufferNode* old_ptr
+        ptr = self.header_buffer
+        while ptr != NULL:
+            old_ptr = ptr
+            ptr = ptr.next
+            free(old_ptr)
+        ptr = self.body_buffer
+        while ptr != NULL:
+            old_ptr = ptr
+            ptr = ptr.next
+            free(old_ptr)
+        # Dealloc curl
+
+    @staticmethod
+    cdef _Response make(Session session, CURL* curl, object future, unsigned long time, Request request):
+        cdef _Response r = _Response.__new__(_Response)
```

Comment:
note to self: document why we do it this way rather than in `__cinit__` (arg types...)

Reply:
done